### ISSUE TITLE
Refactor the parameters for Stock

### ISSUE THREAD

**Issue by sno036** [APSIMInitiative/ApsimX#4617]
So that they are APSIM-like, settable from Managers, and in the JSON file

**Comment by zur003** [comment]
This would require some fairly significant refactoring. It may also mean that GrassGro and AusFarm users with customized parameter sets would not be able to use those parameter sets directly in Apsim simulations. I think we need to consider what we're trying to achieve with this, and decide whether a less radical change would allow us to accommodate most use cases.

**Comment by her123** [comment]
Testing by @sno036 is suggesting that dairy breeds may also be putting too much of their intake into body growth. This may need examination of the parameter.

**Comment by her123** [comment]
First step is to enable the use of external parameter files. By using the grazplan parameter editor it will be simple to edit these data sets and use them in Experiments.
Discussion of improved handling of animal/pasture/crop parameters needs to take place and then Stock could use a new technique for parameter storage.

**Comment by hol353** [comment]
AgResearch are funding my time to refactor STOCK, in particular the parameter handling code and what the user sees in the user interface. I agree this needs to be done carefully and will require some significant refactoring. We think it can be achieved though in the next few months.

1. I have been working on this for a couple of weeks and have come up with a way to read the ruminant.prm file into an AnimalParamSet. This will allow continued use of GrazPlan .prm files into the future solving the compatibility issue. It would be nice to be able to go the other way around as well i.e. take a user genotype parameterisation from APSIM and have it run in GrazPlan. This is a much lower priority for us at the moment.

2. My plan is to allow instances of AnimalParamSet classes to be put onto the STOCK model as a child in the user interface. The user can then change the parameters of these either directly by clicking on the parameter set and modifying parameters or by manager scripts or by factors in an experiment. I currently have this working. There are a lot of parameters so some grouping will probably be desirable at some point.

3. I also want to allow new cross breeds to be specified in the GUI by letting the user add child instances of SingleGenotypeInits to the STOCK model. This will be consistent with the idea of adding AnimalParamSet to the STOCK model. This means we can remove the 'Genotypes' tab on the STOCK view.

4. I then want to do the same for animal groups i.e. let the user add an AnimalInit instance to the STOCK model. This means we can then remove the 'animals' tab on the STOCK view so STOCK in the GUI won't have any parameters at all. Instead it will rely on child models for its parameterisation.

To achieve the above I will raise a number of pull requests in the next few weeks / months. This will provide an opportunity for peer-review as I progress and allow testing of the above by others. 

To make this go smoothly for me can you please refrain from making changes to the STOCK views, presenters and parameter reading / plumbing code, at least for the next few months.

Finally I would welcome any suggestions on the above and any feedback on my pull requests.

**Comment by her123** [comment]
It may be useful for me to merge the changes first in the DairyMilk branch that I made for @sno036 to improve the milk production estimate for dairy breeds.

**Comment by her123** [comment]
@hol353 I have a few ideas for the structure of the genotype and animal child nodes that I need to run past you.

**Comment by hol353** [comment]
Go ahead and get your changes in. Happy to discuss your ideas. If possible would be good if you could outline them on GitHub so that everyone can see them. If not give me a call on my CSIRO work number. I have Jabber running at home.

**Comment by her123** [comment]
Suggestions for adding a Genotype instance child (in gui) to the Stock component:
- a genotype child could be named something like StockGenotype that contains one instance of a SingleGenotypeInits. This is used to populate the Stock.genotypeInits[].
- a StockGenotype object uses the parameter set defined in Stock. code e.g.:
        ```
// Genotype names that can be used in the property presenter
        public List<string> GetGenotypeNames()
        {
            string[] names = new string[0];
            IModel model = Apsim.Parent(this.Parent, typeof(Stock));
            Stock stock = model as Stock;
            if (stock != null)
            {
                names = stock.GenotypeNamesAll();
            }
            return new List<string> (names); 
        }
```
- the PropertyPresenter then uses some code a bit like this:
```
else if (properties[i].Display != null &&
                         properties[i].Display.Type == DisplayType.GenotypeName)
                {
                    cell.EditorType = EditorTypeEnum.DropDown;
                    string[] names = this.getGenotypeNames();
                   ......
```
- need to decide how to specialise the view of properties for cattle/sheep in the property presenter
- each StockGenotype object (in the gui) can contain children of type StockAnimalGroup which each contain an instance of AnimalInits which are used to populate Stock.animalInits[].
- the StockAnimalGroup will need to know if it's parent is cattle or sheep to specialise the property presenter.

**Comment by sno036** [comment]
@her123  - apologies I forgot to get back to you on that. 

I tested the dairy milk change that you made a while back ('before times'!) and it resulted in milk production falling off very quickly (like the cow was drying off) after calving. 

There is a possibility that there was another parameter that should have been changed in the simulation or parameter file? I did the test by running APSIM as it was at that point in time, checkpointed the outputs, pulled your branch with the modification (or was it replacing source files?) and comparing. It didn't seem to be pushing things in the right direction so I backed the changes out. 

Again I might have missed something in the implementation. So you have a before and after?
